CONTRIBUTING.md

```diff
@@ -209,6 +209,72 @@ Before submitting your changes, run the test suite to make sure that nothing is
 ./gradlew check
 ```
 
+### Project layout
+
+This repository is split into many top level directories. The most important
+ones are:
+* docs: Documentation for the project.
+* distribution:  Builds our tar and zip archives and our rpm and deb packages.
+* libs: Libraries used to build other parts of the project. These are meant to
+        be internal rather than general purpose. We have no plans to
+        [semver](https://semver.org/) their APIs or accept feature requests
+        for them. We publish them to maven central because they are
+        dependencies of our plugin test framework, high level rest client, and
+        jdbc driver but.
+* modules:  Features that are shipped with Elasticsearch by default but are not
+        built in to the server. We typically separate features from the server
+        because they require permissions that we don't believe *all* of
+        Elasticsearch should have or because they depend on libraries that we
+        don't believe *all* of Elasticsearch should depend on. For example,
+        reindex requires the `connect` permission so it can perform
```

Comment:
The examples are interesting, but maybe not that important for an overview. Just my 10 cents.